src/curl_async.c

```diff
@@ -549,6 +565,9 @@ int swupd_curl_parallel_download_end(void *handle, int *num_downloads)
 				list_free_item(l, NULL);
 				l = next;
 
+				// Retry was probably scheduled because of network problems, so
+				// reevaluate the number of parallel downloads
+				reevaluate_number_of_parallel_downloads(h, file->retries);
```

Comment:
I'm just getting familiar with the download code, so I may be wrong here.

It looks like there can be a list of failed downloads that can each call reevaluate_number_of_parallel_downloads, even though they failed with the same MAX_XFER value. This would allow a few download failures to drop the MAX_XFER value from 27 -> 1 instead of incrementally.

Reply:
In reevaluate_number_of_parallel_downloads I save the last_retry, so we won't decrement the max_xfer in all calls. Only when the first download is retried by last_retry+1 times.

As we remove and add files to the failed list while we are trying to download, there will be files in the failed list in different retry numbers. Setting a last_retry variable was the only way I could think of on how to avoid the problem you commented.